pkg/streamingpromql/functions/rate.go

```diff
@@ -0,0 +1,176 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/functions.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package functions
+
+import (
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+func Rate(step types.RangeVectorStepData, rangeSeconds float64, floatBuffer *types.FPointRingBuffer, histogramBuffer *types.HPointRingBuffer) (bool, float64, *histogram.FloatHistogram, error) {
+	var err error
+	// Floats
+	fHead, fTail := floatBuffer.UnsafePoints(step.RangeEnd)
+	fCount := len(fHead) + len(fTail)
+
+	// Histograms
+	hHead, hTail := histogramBuffer.UnsafePoints(step.RangeEnd)
+	hCount := len(hHead) + len(hTail)
+
+	if fCount > 0 && hCount > 0 {
+		// We need either at least two Histograms and no Floats, or at least two
+		// Floats and no Histograms to calculate a rate. Otherwise, drop this
+		// Vector element.
+		return false, 0, nil, nil
+	}
+
+	if fCount >= 2 {
```

Comment:
[nit] This `Rate(...)` function is getting quite long - perhaps the bodies of this `if` and the following `if` could be split out to their own functions?